reference/src/glossary.md

```diff
@@ -146,6 +146,11 @@ Moreover, such unsafe code must not return a non-UTF-8 string to the "outside" o
 To summarize: *Data must always be valid, but it only must be safe in safe code.*
 For some more information, see [this blog post](https://www.ralfj.de/blog/2018/08/22/two-kinds-of-invariants.html).
 
+#### Soundness (of code / of a library)
```

Comment:
Should this be structured like so?:

### Soundness

#### Soundness of a type system

#### Soundness of code / a library

### Completeness

Reply:
Soundness of Code is something that every single Rust programmer who ever writes `unsafe {` must know about.

Soundness of Type System... I don't think I even understand this concept fully, and I probably don't ever _need_ to understand it because I don't edit the type system of Rust. Same thing for Completeness. I honestly don't know what that word refers to, but I probably never need to know unless I'm writing an RFC about Rust's type system itself.

I don't think that Soundness of Type System / Completeness have no place here, but they're definitely the secondary usages.

Also, a mention that it is planned to sort all Glossary entries alphabetically once the churn dies down a bit, so we should not design any entry with the assumption that you'll see any other entry right after it.

Reply:
@Centril  I'd rather not turn this into a copy of [nlab](https://ncatlab.org/nlab/show/HomePage).

Reply:
@RalfJung The dense category theory you find in nlab seems quite far apart from a few words *in plain English* about what soundness in general means. I would at least say something like:

> Assuming that undefined behavior does not arise from misuse of `unsafe` in safe code, the typesystem of Rust is sound if no safe code is accepted by it that can cause undefined behavior.

@Lokathor I think it is clear that the purpose of this repo is more than just unsafe code guidelines at this point. (E.g. discussing the RAM more generally.)

Reply:
That's not soundness in general though, that's soundness of a particular type system. It is also standard terminology and not terribly relevant to most Rust programmers (unless they want to change the Rust type system).

Also RAM is a really bad acronym I feel. It's already taken.^^

Reply:
> That's not soundness in general though, that's soundness of a particular type system.

You wanted it to be more Rust-specific, yes? :)

> It is also standard terminology

Sure, but so is really the understanding in the snippet re. usage of `unsafe` -- it's not unique to Rust and applies to any "believe_me" construct you may find in e.g. Agda ([postulates](https://agda.readthedocs.io/en/v2.5.4/language/postulates.html)), Idris ([`believe_me`](https://github.com/idris-lang/Idris-dev/blob/master/libs/prelude/Builtins.idr#L183)), Coq ([axioms](https://github.com/coq/coq/wiki/CoqAndAxioms)), Haskell ([`unsafeCoerce :: a -> b`](https://hackage.haskell.org/package/base-4.12.0.0/docs/Unsafe-Coerce.html#v:unsafeCoerce)), etc.

> and not terribly relevant to most Rust programmers (unless they want to change the Rust type system).

Or if you want to change the R-AM in ways incompatible with the type system rendering it unsound. -- As for what is relevant to "most Rust programmers", I feel there's going to be plenty of stuff in the reference / UCG docs that most Rust devs don't have to care about (thankfully, as you can write large libraries without a single line of `unsafe`).

> Also RAM is a really bad acronym I feel. It's already taken.^^

R-AM? ;)

Reply:
> Sure, but so is really the understanding in the snippet re. usage of unsafe -- it's not unique to Rust and applies to any "believe_me" construct you may find in e.g. Agda (postulates), Idris (believe_me), Coq (axioms), Haskell (unsafeCoerce :: a -> b), etc.

Except that even researches very well-versed in these languages are regularly confused by Rust's `unsafe`. No other language really takes it that serious.

A common reaction to Stacked Borrows is "wait, if the problem is only unsafe code, then what's the problem -- just ignore unsafe code, it that code breaks stuff it's their fault".

If we define soundness of a type system, we have to define *tons* of standard PL terminology. It will make the Rust-specific terminology just drown. If you want to start a second "PL glossary for the working Rust programmer", fine, but I think this glossary here should have a a focus on Rust-specific *concepts* (as opposed to instances of well-known generic concepts).

Reply:
> As for what is relevant to "most Rust programmers", I feel there's going to be plenty of stuff in the reference / UCG docs that most Rust devs don't have to care about (thankfully, as you can write large libraries without a single line of unsafe).

Fair. I meant "most unsafe Rust programmers". The ones for whom this WG is working.

Reply:
@Centril 

We use the word "sound" in a couple of places when talking about the safety invariant that safe Rust wrappers over unsafe Rust code should maintain. The definition provided clarifies those uses.

> Should this be structured like so?:

What problems would doing this solve?

Reply:
@gnzlbg 
> safety invariant 

Good point! I added a link to this new term from the safety invariant description.

Reply:
> Except that even researchers very well-versed in these languages are regularly confused by Rust's `unsafe`. No other language really takes it that serious.
> 
> For example, a common reaction to Stacked Borrows is "wait, if the problem is only unsafe code, then what's the problem -- just ignore unsafe code, if that code breaks stuff it's their fault". That's e.g. how Haskell treats `unsafePerformIO`. Similarly, OCaml also makes no effort to even spell out when and how `Obj.magic` is correct to use.

I think this is primarily a consequence of `unsafePerformIO` being relatively rare in comparison to `unsafe` in Rust. That seems expected. However, as one sees from [the Coq page on `Axiom`s](https://github.com/coq/coq/wiki/CoqAndAxioms) it does seem to be taken seriously when the soundness of proofs depends on it.

> If we define soundness of a type system, we have to define _tons_ of standard PL terminology. It will make the Rust-specific terminology just drown in the "noise". If you want to start a second "PL glossary for the working Rust programmer", fine, but I think this glossary here should have a a focus on Rust-specific _concepts_ (as opposed to instances of well-known generic concepts).

I am content to just have something like that one sentence I wrote about what soundness for Rust's type system means; I think it gives an intuition for soundness overall without going into "tons of standard PL terminology". I do not see how that would follow -- it feels like a strawman?

> What problems would doing this solve?

It would help to give an intuition re. ["Unsafe as a Human-Assisted Type System"](https://matklad.github.io/2019/07/25/unsafe-as-a-type-system.html) grounding `unsafe` as a mechanism for "User-land ways to extend the type system".

Reply:
> I think this is primarily a consequence of unsafePerformIO being relatively rare in comparison to unsafe in Rust. That seems expected. However, as one sees from the Coq page on Axioms it does seem to be taken seriously when the soundness of proofs depends on it.

Proof assistants are very serious about this, that's true, but they also typically don't let axioms cause UB -- if computation happens, they only get stuck.

I have heard contradicting things about `unsafePerformIO` usage in Haskell; about half the Haskell people I meet say its use is not uncommon and you'll find it in production-grade libraries, others say it should never be used in production code.

`Obj.magic` in OCaml... well, it's used by Coq, for example. ;)

Reply:
@Centril I added some context. I found that hard to do without defining "Undefined Behavior" so I did that, too.

Reply:
> I have heard contradicting things about `unsafePerformIO` usage in Haskell; about half the Haskell people I meet say its use is not uncommon and you'll find it in production-grade libraries, others say it should never be used in production code.

As far as I know `unsafePerformIO` and friends are actually used in the base library so it's hard to not transitively depend on it... I venture *a guess* that the more eggheaded users of Haskell would dislike the footguns more than some production users. ^^ Some links [for](https://www.schoolofhaskell.com/user/snoyberg/general-haskell/advanced/evaluation-order-and-state-tokens#unsafeperformio--unsafedupableperformio--inlineperformio--aka-he-who-shall-not-be-named-) the [curious](https://stackoverflow.com/questions/10529284/is-there-ever-a-good-reason-to-use-unsafeperformio?rq=1).

> @Centril I added some context. I found that hard to do without defining "Undefined Behavior" so I did that, too.

This turned out even better than what I wanted. ❤️